manager/dispatcher/orphans.go

```diff
@@ -0,0 +1,155 @@
+package dispatcher
+
+import (
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+// downNodeStore defines a type to store the nodes which are currently
+// down.
+type downNodeStore struct {
+	// downNodeCheckInterval specifies the frequency at which the
+	// down nodes are scanned.
+	downNodeCheckInterval time.Duration
+
+	// downNodeOrphanThreshold specifies the how long before the
+	// tasks that belong to down nodes are moved to ORPHANED
+	// state.
+	downNodeOrphanThreshold time.Duration
+
+	// nodes is a map of nodes with nodeID as the key and scan
+	// count which tracks the timer expiry as the value.
+	nodes map[string]int
+	store *store.MemoryStore
+	sync.Mutex
+}
+
+func newDownNodeStore(s *store.MemoryStore, downNodeCheckInterval, downNodeOrphanThreshold time.Duration) *downNodeStore {
+	dn := &downNodeStore{
+		downNodeCheckInterval:   downNodeCheckInterval,
+		downNodeOrphanThreshold: downNodeOrphanThreshold,
+		nodes: make(map[string]int),
+		store: s,
+	}
+
+	s.View(func(readTx store.ReadTx) {
```

Comment:
I think I may have been wrong when I suggested adding this :(. I believe the dispatcher sets all nodes to `UNKNOWN` when it starts up, and then any nodes that it doesn't hear from become `DOWN` shortly thereafter. So it seems like any nodes which are still down would be passed to `add`.